# TP4 Chatbots

# Chatbots basados en recuperación

En inglés information retrieval chatbots

# Esteban Matias Cancino

# Motor de búsqueda

* Búsqueda por palabras clave: Extrae palabras clave de la pregunta del usuario y busca coincidencias en las preguntas almacenadas.

* Similitud del coseno: Si has representado las preguntas como vectores (por ejemplo, usando TF-IDF o word embeddings), puedes usar la similitud del coseno para medir la distancia entre las preguntas.

* Word embeddings: Utiliza modelos de word embeddings como Word2Vec o BERT para obtener representaciones semánticas de las preguntas y las consultas del usuario.

### Búsqueda por palabras claves

In [ ]:
# tu_diccionario = {
#    "hola": "¡Hola! ¿En qué puedo ayudarte?",
#    "adiós": "Hasta luego. ¡Que tengas un buen día!",
#    "información": "¿Qué tipo de información estás buscando?",
#    # Agrega más entradas de diccionario según tus necesidades
# }


In [ ]:
# import spacy
# nlp = spacy.load("es_core_news_md")
# def responder_pregunta(pregunta):
#     pregunta_procesada = nlp(pregunta.lower())  # Procesa la pregunta y convierte a minúsculas
#     respuesta = "Lo siento, no entiendo tu pregunta."

#     # Busca una coincidencia en el diccionario
#     for palabra in pregunta_procesada:
#         # regresa la primer coincidencia que encuentra
#         if palabra.text in tu_diccionario:
#             respuesta = tu_diccionario[palabra.text]
#             break

#     return respuesta


In [ ]:
# while True:
#     entrada_usuario = input("Tú: ")
#     if entrada_usuario.lower() == "salir":
#         print("Chatbot: Hasta luego.")
#         break
#     respuesta = responder_pregunta(entrada_usuario)
#     print("Chatbot:", respuesta)


Puede modificar la implementación anterior para evitar el bucle while True y usar un dataset de prueba de preguntas y respuestas.

## Búsqueda por similitud

Para los chatbots basados ​​en recuperación, es común utilizar bolsas de palabras (bag of words) o tf-idf para calcular la similitud de intenciones.

In [ ]:
# from sklearn.feature_extraction.text import TfidfVectorizer
# from sklearn.metrics.pairwise import cosine_similarity

# # Datos de ejemplo
# preguntas = ["¿Qué es el aprendizaje automático?",
#              "¿Cómo funciona la regresión lineal?"]
# respuestas = ["El aprendizaje automático es una rama de la inteligencia artificial...",
#               "La regresión lineal es un método de modelado..."]

# # Vectorización con TF-IDF
# vectorizer = TfidfVectorizer()
# tfidf_matrix = vectorizer.fit_transform(preguntas)

# # Función para encontrar la mejor coincidencia
# def responder_pregunta(consulta_usuario):
#     consulta_vec = vectorizer.transform([consulta_usuario])
#     similitudes = cosine_similarity(consulta_vec, tfidf_matrix).flatten()
#     print(similitudes)
#     indice_mejor_coincidencia = similitudes.argmax()
#     print(indice_mejor_coincidencia)
#     return respuestas[indice_mejor_coincidencia]


In [ ]:

# # Ejemplo de consulta
# consulta = "¿Qué es la regresión lineal?"
# print(responder_pregunta(consulta))


## Búsqueda por similitud en embeddings

Puedes vectorizar el texto usando embeddings, de spacy por ejemplo, como vimos en clases.


In [ ]:
# !pip install spacy --quiet
# !python -m spacy download es_core_news_sm --quiet


## Actividades



### 1) Elaborar un dataset de preguntas y respuestas para crear un Chatbot para un aplicación particular. ( 3 puntos )

1.1 Debe definir la aplicación (atención al cliente bancario, atención a estudiantes universitarios, etc).

1.2 El listado de preguntas y respuestas debe tener como mínimo 20 elementos pregunta - respuesta.

###  2) Crear el chatbot utilizando TFIDF y similitud del coseno. (1 punto)

### 3) Crear otro chatbot utilizando embeddings. Indique cuál embedding (1 punto) pre-entrenado eligió.

### 4) Muestra ambos chatbots funcionando (1 punto)

Adjuntar la lista de preguntas y respuestas utilizadas para probar el funcionamiento.

Releve o indique cuáles respondió correctamente y cuáles no.

### 5) Añade tus conclusiones de todo lo realizado (2 punto)

* Resalte e indique en cuáles respuestas falla o tiene problemas.
* Cuáles preguntas confunde.
* Compare los resultados de los chatbots.



### No olvides:

* Explicar tus decisiones y configuraciones. Añadir tus conclusiones.
* Anunciar en el foro cuál será tu aplicación y postear tu entrega y tus avances.
* Debes subir tu notebook a un repo GitHub público de tu propiedad compartido + enlace colab.
* Documentar todo el proceso.
* Citar tus fuentes





**DEADLINE ENTREGA HASTA PUNTO 5 PARA REGULARIZAR O PROMOCIÓN** DOMINGO 09/11

## Elaborar un dataset de preguntas y respuestas para crear un Chatbot para un aplicación particular

### Definición de la aplicación

Elegí como aplicación la atención a estudiantes universitarios

### Listado de preguntas y respuestas

In [ ]:
preguntas_respuestas = {
  "preguntas" : [
    "hola", "adiós",
    "¿Cuáles son los requisitos para la admisión?",
    "¿Cómo me inscribo en la universidad?",
    "¿Qué carreras ofrece la Universidad Tech?",
    "¿Cuánto cuesta la matrícula por semestre?",
    "¿Hay becas disponibles para estudiantes?",
    "¿Cómo solicito una beca?",
    "¿Cuáles son los horarios de clases?",
    "¿Dónde encuentro el calendario académico?",
    "¿Cómo cambio mi contraseña del portal estudiantil?",
    "¿Qué servicios de salud ofrece la universidad?",
    "¿Hay residencias universitarias?",
    "¿Cómo me uno a un club estudiantil?",
    "¿Cuándo son los exámenes finales?",
    "¿Cómo preparo mi tesis?",
    "¿Qué es el sistema de créditos?",
    "¿Puedo convalidar materias de otra universidad?",
    "¿Cómo contacto al decano?",
    "¿Hay programas de intercambio internacional?",
    "¿Qué hago si pierdo mi carnet estudiantil?",
    "¿Cuáles son las fechas de inscripción a materias?",
    "¿Cómo accedo a la biblioteca virtual?",
    "¿Hay tutorías disponibles?",
    "¿Qué deportes se practican en la universidad?",
    "¿Cómo reporto un problema en el campus?",
    "¿Cuáles son las normas de convivencia?",
    "¿Cómo me inscribo a las materias?",
    "¿Dónde puedo ver mi horario de clases?",
    "¿Cómo solicito un certificado de alumno regular?",
    "¿Dónde está la biblioteca?",
    "¿Cuál es el horario de atención de secretaría?",
    "¿Cómo consulto mis notas?",
    "¿Qué requisitos necesito para inscribirme a una materia?",
    "¿Cuándo empiezan las clases?",
    "¿Cómo puedo cambiar mi contraseña del campus virtual?",
    "¿Dónde pago la matrícula?",
    "¿Hay becas disponibles?",
    "¿Cómo solicito una beca estudiantil?",
    "¿Cuál es el plan de estudios de mi carrera?",
    "¿Puedo cursar materias de otras carreras?",
    "¿Cómo justifico una inasistencia?",
    "¿Cuántas faltas puedo tener?",
    "¿Dónde está el comedor universitario?",
    "¿Hay estacionamiento para estudiantes?",
    "¿Cómo contacto a mi profesor?",
    "¿Qué hago si pierdo mi credencial estudiantil?",
    "¿Cuándo es el período de inscripción?",
    "¿Puedo cursar materias en contraturno?",
    "¿Cómo tramito mi título universitario?",
    "¿Cómo presento un reclamo o queja formal?",
    "¿Hay apoyo para estudiantes con discapacidad?"],

  "respuestas" : [
    "¡Hola! ¿En qué puedo ayudarte?", "Hasta luego. ¡Que tengas un excelente día!",
    "Los requisitos incluyen título secundario, examen de ingreso y documentos personales. Visita el sitio web para detalles.",
    "La inscripción se hace online a través del portal de admisiones. Necesitas crear una cuenta y subir documentos.",
    "Ofrecemos carreras en Ingeniería, Ciencias, Humanidades, Negocios y Artes. Consulta el catálogo completo en el sitio.",
    "La matrícula varía por carrera, pero promedia $5000 por semestre. Hay descuentos para becarios.",
    "Sí, hay becas por mérito, necesidad económica y deportivas. Aplican restricciones.",
    "Solicita becas en la oficina de ayuda financiera con formulario y documentos probatorios.",
    "Los horarios dependen de la carrera, pero clases son de 8am a 8pm. Revisa tu horario personal en el portal.",
    "El calendario académico está disponible en el portal estudiantil, con fechas de inicio, fin y feriados.",
    "Ve al portal, selecciona 'Olvidé contraseña' y sigue las instrucciones por email.",
    "Ofrecemos consultas médicas, psicológicas y dentales gratuitas en el centro de salud del campus.",
    "Sí, hay residencias con habitaciones compartidas. Aplica en la oficina de vivienda.",
    "Visita la feria de clubs al inicio del semestre o contacta a la asociación estudiantil.",
    "Los exámenes finales son en junio y diciembre, según el calendario académico.",
    "Para la tesis, elige un tema, un tutor y sigue las guías del departamento. Hay talleres disponibles.",
    "El sistema de créditos asigna puntos por materia aprobada. Necesitas 180 créditos para graduarte.",
    "Sí, presenta certificados y solicita convalidación en la secretaría académica.",
    "Envía un email a decano@utech.edu o agenda una cita en su oficina.",
    "Sí, programas con universidades en Europa y América. Aplica en la oficina de internacionalización.",
    "Reporta la pérdida en seguridad y solicita un duplicado con foto y pago.",
    "Las inscripciones a materias son del 1 al 15 de cada mes previo al semestre.",
    "Accede a la biblioteca virtual con tu usuario y contraseña del portal estudiantil.",
    "Sí, hay tutorías gratuitas en matemáticas, idiomas y más. Inscríbete en el centro de apoyo.",
    "Practicamos fútbol, básquet, natación y más. Únete a los equipos en el departamento de deportes.",
    "Reporta problemas como mantenimiento o seguridad a través de la app del campus o al teléfono 1234.",
    "Las normas incluyen respeto, no discriminación y cuidado del ambiente. Lee el código de conducta.",
    "Para inscribirte a las materias debes ingresar al campus virtual con tu usuario y contraseña, ir a la sección 'Inscripciones' y seleccionar las materias disponibles según tu plan de estudios. El período de inscripción es del 1 al 15 de cada cuatrimestre.",
    "Puedes ver tu horario de clases en el campus virtual, sección 'Mi Horario', o descargarlo en formato PDF desde 'Descargas'. También lo encontrarás en las carteleras de tu facultad.",
    "Para solicitar un certificado de alumno regular debes ingresar a 'Trámites' en el campus virtual, seleccionar 'Certificado de Alumno Regular' y completar el formulario. El certificado estará disponible en 48 horas hábiles.",
    "La biblioteca central está ubicada en el edificio B, planta baja. El horario de atención es de lunes a viernes de 8:00 a 20:00 hs y sábados de 9:00 a 13:00 hs.",
    "El horario de atención de secretaría académica es de lunes a viernes de 9:00 a 17:00 hs. Los trámites administrativos también pueden realizarse a través del campus virtual las 24 horas.",
    "Puedes consultar tus notas ingresando al campus virtual, sección 'Calificaciones'. Las notas son publicadas por los profesores dentro de los 10 días posteriores al examen o evaluación.",
    "Para inscribirte a una materia necesitas: 1) Tener aprobadas las materias correlativas, 2) Estar al día con los pagos administrativos, 3) No tener sanciones disciplinarias vigentes. Consulta las correlatividades en tu plan de estudios.",
    "Las clases del primer cuatrimestre comienzan la primera semana de marzo, y las del segundo cuatrimestre la primera semana de agosto. Las fechas exactas se publican en el calendario académico anual.",
    "Para cambiar tu contraseña del campus virtual, haz clic en 'Olvidé mi contraseña' en la pantalla de inicio de sesión. Te enviaremos un enlace a tu correo electrónico institucional para restablecer la contraseña.",
    "Puedes pagar la matrícula en: 1) Cajeros automáticos usando el código de pago electrónico, 2) Plataformas de pago online (Mercado Pago, Rapipago, Pago Fácil), 3) Tesorería de la universidad de lunes a viernes de 9:00 a 15:00 hs.",
    "Sí, la universidad ofrece diferentes tipos de becas: becas de mérito académico, becas socioeconómicas, becas deportivas y becas de investigación. Consulta los requisitos y convocatorias en la sección 'Becas y Ayuda Económica' del sitio web.",
    "Para solicitar una beca debes: 1) Completar el formulario online en 'Becas', 2) Adjuntar la documentación requerida (certificado de ingresos, analítico académico), 3) Esperar la evaluación del comité. El proceso demora aproximadamente 30 días hábiles.",
    "El plan de estudios de tu carrera está disponible en el sitio web de la facultad, en la sección 'Carreras'. También puedes consultarlo en secretaría académica o descargarlo desde el campus virtual en formato PDF.",
    "Sí, puedes cursar hasta 2 materias electivas de otras carreras si hay cupos disponibles y cumplen con tu plan de estudios. Debes solicitar autorización en secretaría académica presentando una nota fundamentando tu pedido.",
    "Para justificar una inasistencia debes presentar el certificado médico o documentación correspondiente en secretaría dentro de las 48 horas. Las inasistencias por razones de fuerza mayor se evalúan caso por caso.",
    "El régimen de asistencia establece que puedes tener hasta un 25% de inasistencias (equivale a 3-4 clases según la materia). Superado ese límite, pierdes la regularidad de la materia y debes cursarla nuevamente.",
    "El comedor universitario está ubicado en el edificio C, primer piso. Ofrece menú estudiantil a precio subsidiado. Horarios: desayuno 7:30-9:00, almuerzo 12:00-15:00, merienda 16:00-18:00. Debes presentar tu credencial estudiantil.",
    "Sí, hay estacionamiento para estudiantes en el predio universitario. Es necesario tramitar el permiso de estacionamiento en la oficina de seguridad (edificio A). El costo es de $2000 mensuales y debes presentar cédula verde del vehículo y credencial.",
    "Puedes contactar a tu profesor a través del campus virtual usando la mensajería interna, o por correo electrónico institucional. Los profesores tienen horarios de consulta publicados en sus cátedras, generalmente 1 hora antes o después de cada clase.",
    "Si pierdes tu credencial estudiantil debes: 1) Hacer la denuncia en 'Reportes' del campus virtual, 2) Solicitar el duplicado en oficina de alumnos presentando DNI y pagando la tasa administrativa ($500), 3) Retirar la nueva credencial en 5 días hábiles.",
    "El período de inscripción a materias es: Primer cuatrimestre: del 1 al 15 de febrero. Segundo cuatrimestre: del 15 al 30 de julio. Existe un período extraordinario de inscripción en los primeros 5 días de cada cuatrimestre con recargo.",
    "Sí, puedes cursar materias en contraturno si existen comisiones disponibles y no se superponen con tus otras materias. Debes solicitarlo durante el período de inscripción indicando los motivos (laborales, etc) en el formulario correspondiente.",
    "Para tramitar tu título universitario debes: 1) Haber aprobado todas las materias del plan de estudios, 2) Completar el formulario de solicitud de título, 3) Presentar documentación (DNI, partida de nacimiento, certificado analítico), 4) Pagar la tasa de título. El trámite demora 90-120 días.",
    "Puedes presentar un reclamo o queja formal en 'Atención al Estudiante' completando el formulario online en la sección 'Reclamos' del campus virtual, o personalmente en la oficina de atención al público. Guarda el comprobante y los números de expediente para seguimiento.",
    "Sí, el Centro de Inclusión y Atención a la Diversidad ofrece apoyos para estudiantes con discapacidad: adaptaciones curriculares, tutorías especializadas, intérpretes y asesoramiento. Solicita una evaluación y el plan de apoyos en la oficina correspondiente."]}

print(f"cantidad de preguntas {len(preguntas_respuestas["preguntas"])}, cantidad de respuestas {len(preguntas_respuestas["respuestas"])}")

cantidad de preguntas 52, cantidad de respuestas 52


In [ ]:
# Instalo la librería spaCy
!pip install spacy
# Descargo e instalo el modelo de idioma español 'es_core_news_md' (modelo mediano: buena relación entre tamaño y calidad).
!python -m spacy download es_core_news_md

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 MB 12.8 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_md')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


### cargo las librerias que utilizaré

In [ ]:
# NumPy para operaciones numéricas eficientes y manejo de arreglos
import numpy as np
# pandas para manipulación y análisis de datos mediante DataFrames
import pandas as pd
# Importo el módulo 're' para trabajar con expresiones regulares como búsqueda, reemplazo y limpieza de texto
import re
# Importo unicodedata para normalizar cadenas Unicode y facilitar la remoción de acentos
import unicodedata
# spaCy para procesamiento de lenguaje natural y uso de modelos lingüísticos
import spacy
# TfidfVectorizer para transformar colecciones de documentos en matrices TF-IDF.
from sklearn.feature_extraction.text import TfidfVectorizer
# cosine_similarity para calcular la similitud coseno entre vectores
from sklearn.metrics.pairwise import cosine_similarity

## Crear el chatbot utilizando TFIDF y similitud del coseno

In [ ]:
# Crear DataFrame
df = pd.DataFrame(preguntas_respuestas)
df.head()

preguntas  \
0                                          hola   
1                                         adiós   
2  ¿Cuáles son los requisitos para la admisión?   
3          ¿Cómo me inscribo en la universidad?   
4     ¿Qué carreras ofrece la Universidad Tech?   

                                          respuestas  
0                     ¡Hola! ¿En qué puedo ayudarte?  
1         Hasta luego. ¡Que tengas un excelente día!  
2  Los requisitos incluyen título secundario, exa...  
3  La inscripción se hace online a través del por...  
4  Ofrecemos carreras en Ingeniería, Ciencias, Hu...

In [ ]:
# Defino una función para limpiar y normalizar texto (quita nulos, mayúsculas, acentos, puntuación y espacios extra)
def limpiar_texto(texto):
    # Si el valor es NaN lo convierto a cadena vacía para evitar errores posteriores
    if pd.isna(texto): return ""
    # Convierto todo a minúsculas para homogeneizar comparaciones y búsquedas
    texto = texto.lower()
    # quito los acentos:
    # - Normalizo la cadena a forma NFD (descompone caracteres compuestos en base + marcas)
    # - Elimino los caracteres cuya categoría Unicode es 'Mn' (marca, como tilde o acento)
    texto = ''.join(c for c in unicodedata.normalize('NFD', texto) if unicodedata.category(c) != 'Mn')
    # elimino cualquier carácter que no sea una letra a-z, un dígito o un espacio
    # Esto quita puntuación, signos de interrogación/exclamación invertidos, símbolos, etc
    texto = re.sub(r'[^a-z0-9\s]', '', texto)
    # Remuevo espacios repetidos y espacios al inicio/fin dejando una sola separación entre palabras
    texto = ' '.join(texto.split())
    # Devuelvo el texto limpio
    return texto

# Aplico la función de limpieza a la columna 'preguntas' del DataFrame
df['preguntas'] = df['preguntas'].apply(limpiar_texto)

# Muestro las primeras filas del DataFrame
df.head()

preguntas  \
0                                        hola   
1                                       adios   
2  cuales son los requisitos para la admision   
3          como me inscribo en la universidad   
4     que carreras ofrece la universidad tech   

                                          respuestas  
0                     ¡Hola! ¿En qué puedo ayudarte?  
1         Hasta luego. ¡Que tengas un excelente día!  
2  Los requisitos incluyen título secundario, exa...  
3  La inscripción se hace online a través del por...  
4  Ofrecemos carreras en Ingeniería, Ciencias, Hu...

In [ ]:
# inicializo el vectorizador TF-IDF
vectorizer = TfidfVectorizer()

# ajusto el vectorizador con todas las preguntas del DataFrame y transformo ese texto en una matriz TF-IDF
tfidf_matrix = vectorizer.fit_transform(df['preguntas'])

In [ ]:
# defino la función del chatbot que recibe una consulta y devuelve la respuesta más similar junto con la similitud
def chatbot_tfidf(consulta):
    # limpio y normalizo la consulta con la función de preprocesamiento definida anteriormente
    consulta_limpia = limpiar_texto(consulta)
    # transformo la consulta limpia a un vector TF-IDF usando el vectorizador ya ajustado
    consulta_vec = vectorizer.transform([consulta_limpia])
    # calculo la similitud coseno entre la consulta y todas las preguntas del corpus
    similitudes = cosine_similarity(consulta_vec, tfidf_matrix).flatten()
    # extraigo la similitud máxima para evaluar si existe una coincidencia suficientemente buena
    max_sim = max(similitudes)
    # si la similitud máxima está por debajo del umbral, devuelvo un mensaje de no comprensión y la puntuación
    if max_sim < 0.5: return "Lo siento, no entiendo tu pregunta. ¿Puedes reformularla?", max_sim
    # obtengo el índice de la pregunta con mayor similitud
    indice = np.argmax(similitudes)
    # devuelvo la respuesta asociada a ese índice y la similitud máxima
    return df["respuestas"][indice], max_sim

In [ ]:
# hago una prueba para ver que tal funcina
respuesta_tfidf, similitud_tfidf = chatbot_tfidf("hola")
print(f"Chatbot TF-IDF: {respuesta_tfidf} (Similitud: {similitud_tfidf:.4f})")

Chatbot TF-IDF: ¡Hola! ¿En qué puedo ayudarte? (Similitud: 1.0000)


### Crear otro chatbot utilizando embeddings

elegí spaCy porque es ligero y soporta español

In [ ]:
# cargo el modelo de spaCy en español.
nlp = spacy.load("es_core_news_md")

In [ ]:
# genero los embeddings para cada pregunta del DataFrame usando el modelo de spaCy.
# recorro cada pregunta ya limpiada y obtengo su vector semántico
embeddings_preguntas = [nlp(p).vector for p in df["preguntas"]]

In [ ]:
# defino la función del chatbot basado en embeddings, que calcula similitud semántica entre la consulta y las preguntas
def chatbot_embeddings(consulta):
    # limpio la consulta para asegurar consistencia con el preprocesamiento aplicado a las preguntas
    consulta_limpia = limpiar_texto(consulta)
    # obtengo el embedding (vector semántico) de la consulta usando el modelo spaCy
    consulta_emb = nlp(consulta_limpia).vector
    # calculo la similitud coseno entre el vector de la consulta y cada embedding de las preguntas almacenadas
    # para cada embedding, genero una matriz 1x1 con el valor de similitud
    similitudes = [cosine_similarity([consulta_emb], [emb])[0][0] for emb in embeddings_preguntas]
    # obtengo la mayor similitud para evaluar si alguna pregunta coincide suficientemente bien
    max_sim = max(similitudes)
    # si la similitud máxima es menor al umbral definido (0.5), devuelvo un mensaje indicando que no hay coincidencias claras
    if max_sim < 0.5: return "Lo siento, no entiendo tu pregunta. ¿Puedes reformularla?", max_sim
    # identifico el índice de la pregunta con mayor similitud
    indice = np.argmax(similitudes)
    # devuelvo la respuesta correspondiente a esa pregunta junto con la similitud máxima
    return df["respuestas"][indice], max_sim

In [ ]:
# pruebo tambien para ver su funcionamiento
respuesta_embedding, similitud_embedding = chatbot_embeddings("hola como estas")
print(f"Chatbot embedding: {respuesta_embedding} (Similitud: {similitud_embedding:.4f})")

Chatbot embedding: Las clases del primer cuatrimestre comienzan la primera semana de marzo, y las del segundo cuatrimestre la primera semana de agosto. Las fechas exactas se publican en el calendario académico anual. (Similitud: 0.5865)


### Muestra ambos chatbots funcionando

In [ ]:
# defino las preguntas para testear los modelos
preguntas_test = [
    "¿Cuáles son los requisitos para la admisión?",
    "¿Cómo inscribirme en la uni?",
    "¿Becas para alumnos?",
    "¿Horarios de exámenes?",
    "¿Qué es el cambio climático?",
    "¿Cómo cambio contraseña?",
    "¿Programas de intercambio?",
    "¿Normas de la universidad?",
    "¿Dónde está la biblioteca?",
    "¿Fechas para inscribir materias?"]

# defino una función para comparar las respuestas de los dos chatbots (TF-IDF y Embeddings) sobre un conjunto de preguntas
def comparar_bots_df(preguntas_usuario):
    # inicializo una lista donde guardaré los resultados generados por cada pregunta evaluada
    resultados = []
    # recorro cada pregunta de la lista
    for pregunta in preguntas_usuario:
        # limpio la pregunta para mantener coherencia con el preprocesamiento de ambos chatbots
        pregunta_limpia = limpiar_texto(pregunta)
        # obtengo la respuesta del chatbot basado en TF-IDF
        resp_tfidf = chatbot_tfidf(pregunta_limpia)
        # obtengo la respuesta del chatbot basado en embeddings semánticos
        resp_emb = chatbot_embeddings(pregunta_limpia)
        # agrego los resultados en forma de diccionario para construir el DataFrame
        resultados.append({
            "Pregunta": pregunta,
            "Respuesta TF-IDF": resp_tfidf,
            "Respuesta Embeddings": resp_emb})

    # convierto la lista de resultados en un DataFrame para visualizar y analizar comparativamente ambas respuestas
    df_resultados = pd.DataFrame(resultados)
    # configuro pandas para mostrar el contenido completo de las columnas sin recortes
    pd.set_option('display.max_colwidth', None)
    # devuelvo el DataFrame con el resumen comparativo
    return df_resultados

In [ ]:
# ejecuto la función de comparación usando el conjunto de preguntas de prueba y guardo el resultado en un DataFrame
df_comparacion = comparar_bots_df(preguntas_test)

# muestro el DataFrame resultante con las respuestas generadas por ambos chatbots
display(df_comparacion)

Pregunta  \
0  ¿Cuáles son los requisitos para la admisión?   
1                  ¿Cómo inscribirme en la uni?   
2                          ¿Becas para alumnos?   
3                        ¿Horarios de exámenes?   
4                  ¿Qué es el cambio climático?   
5                      ¿Cómo cambio contraseña?   
6                    ¿Programas de intercambio?   
7                    ¿Normas de la universidad?   
8                    ¿Dónde está la biblioteca?   
9              ¿Fechas para inscribir materias?   

                                                                                                                                                         Respuesta TF-IDF  \
0                          (Los requisitos incluyen título secundario, examen de ingreso y documentos personales. Visita el sitio web para detalles., 1.0000000000000002)   
1                                                                                        (Lo siento, no entiendo tu pregunta. ¿Puedes reformularla?, 0.45570445354578404)   
2                                                                (Sí, hay becas por mérito, necesidad económica y deportivas. Aplican restricciones., 0.6513810968317382)   
3                                                                                         (Lo siento, no entiendo tu pregunta. ¿Puedes reformularla?, 0.4349727213710096)   
4                                                                                         (Lo siento, no entiendo tu pregunta. ¿Puedes reformularla?, 0.4918685654111643)   
5                                                                 (Ve al portal, selecciona 'Olvidé contraseña' y sigue las instrucciones por email., 0.6533260769819372)   
6                                                (Sí, programas con universidades en Europa y América. Aplica en la oficina de internacionalización., 0.7822518230130686)   
7                                                                                        (Lo siento, no entiendo tu pregunta. ¿Puedes reformularla?, 0.42898936486579425)   
8  (La biblioteca central está ubicada en el edificio B, planta baja. El horario de atención es de lunes a viernes de 8:00 a 20:00 hs y sábados de 9:00 a 13:00 hs., 1.0)   
9                                                                      (Las inscripciones a materias son del 1 al 15 de cada mes previo al semestre., 0.5069001019618596)   

                                                                                                                                                                                                                         Respuesta Embeddings  
0                                                                                                       (Los requisitos incluyen título secundario, examen de ingreso y documentos personales. Visita el sitio web para detalles., 1.0000001)  
1                                                               (La biblioteca central está ubicada en el edificio B, planta baja. El horario de atención es de lunes a viernes de 8:00 a 20:00 hs y sábados de 9:00 a 13:00 hs., 0.78550094)  
2                                                                                                                                            (Sí, hay becas por mérito, necesidad económica y deportivas. Aplican restricciones., 0.79707915)  
3                                                                                                                             (Sí, programas con universidades en Europa y América. Aplica en la oficina de internacionalización., 0.7589258)  
4                                                                                                                             (El sistema de créditos asigna puntos por materia aprobada. Necesitas 180 créditos para graduarte., 0.86249655)  
5          (Para justificar una inasistencia debes presentar el certificado médico o documentación correspondiente en secretaría dent

Los resultados de similitud de respuesta para cada chatbot fueron:

0. TF-IDF = 1.00 │ Embeddings = 1.00

1. TF-IDF = 0.46 │ Embeddings = 0.78

2. TF-IDF = 0.65 │ Embeddings = 0.80

3. TF-IDF = 0.43 │ Embeddings = 0.76

4. TF-IDF = 0.49 │ Embeddings = 0.86

5. TF-IDF = 0.65 │ Embeddings = 0.72

6. TF-IDF = 0.78 │ Embeddings = 0.88

7. TF-IDF = 0.43 │ Embeddings = 0.86

8. TF-IDF = 1.00 │ Embeddings = 1.00

9. TF-IDF = 0.51 │ Embeddings = 0.70




TF-IDF respondió correctamente en los índices: 0, 2, 5, 6, 8, 9.
(Correctas: requisitos de admisión; información sobre becas; cambio de contraseña; programas de intercambio; ubicación de la biblioteca; fechas de inscripción.)

Embeddings tomando encuenta la coherencia de la respuesta y no el umbral, el modelo respondió correctamente en los índices: 0, 2, 6, 8.
(Correctas: requisitos de admisión; becas; programas de intercambio; ubicación de la biblioteca.)

### Añade tus conclusiones de todo lo realizado

#### Análisis de Fallos

**TF-IDF:**
- Falló en preguntas 1, 3, 4, 7 por falta de palabras clave exactas
- Ejemplo: "¿Cómo inscribirme en la uni?" → la palabra "uni" no existe en el corpus
- Es muy sensible a variaciones léxicas y abreviaturas

**Embeddings:**
- Fallos más graves: dio respuestas incorrectas con alta confianza
- Pregunta 1 (Cómo inscribirme en la uni): respondió sobre la biblioteca central (0.78)
- Pregunta 3 (horarios exámenes): respondió sobre intercambio (0.76)
  → Confusión por contexto universitario general
- Pregunta 4 (cambio climático): respondió sobre créditos académicos (0.86)
  → Esto indica que el modelo encuentra similitud semántica superficial
- Pregunta 5 (¿Cómo cambio contraseña?): respondió sobre justificación de inasistencias y certificados médicos (0.72).
  → Respuesta fuera de contexto, confusión por términos administrativos/portal.
- Pregunta 7 (¿Normas de la universidad?): respondió sobre servicios médicos/psicológicos del campus (0.86).
→ Confusión por vocabulario compartido de servicios del campus.
- Pregunta 9 (Fechas para inscribir materias): respondió sobre cursada de materias (0.70)
  → Confusión por el uso de la palabra materias?
#### Confusiones Identificadas

**TF-IDF confunde:**
- Sinónimos y abreviaturas ("uni" vs "universidad")
- Reformulaciones de preguntas

**Embeddings confunde:**
- Temas relacionados pero diferentes
- "Cambio climático" con "sistema de créditos" (quizas tienen una similitud matematica?)

**problema que note en ambos modelos**
algo que note haciendo pruebas con ambos modelos es que no logran obtener una buena similitud con palabras con y sin acentos

#### Comparación Final

| Métrica | TF-IDF | Embeddings | Ganador |
|---------|--------|------------|---------|
| Precisión | 6/10 (60%) | 4/10 (40%) | TF-IDF |
| Falsos positivos | Bajo | Alto | TF-IDF |


**CONCLUSIÓN PRINCIPAL:**
Para este chatbot universitario, TF-IDF funciona mejor porque:
1. el Dataset es pequeño (52 preguntas) y funciona mejor con datasets pequeños, el modelo de embeddings quizas necesitaria un dataset mas amplio
2. Mayor precisión (menos errores graves)

Embeddings sería mejor con:
1. quizas con un Dataset más grande
2. Ajustar del umbral subirlo a 0.85-0.90 para obtener las respuestas mas precisas y ver si se evita la correlacion de respuestas con preguntas que no tienen relacion

### Fuentes
Apuntes de clase proporcionados por el docente de la materia.

Asistencia de ChatGPT (modelo de lenguaje de OpenAI) para la generacion de la base de datos.

# TP5 - Chatbots con RAG (Retrieval-Augmented Generation)

## 1) CONSIGNAS A RESOLVER

### a) Creación del conjunto de datos de evaluación.
Además del dataset original que creó, debe crear un dataset de prueba o evaluación con la misma lógica: preguntas y respuestas.

### b) Debe elegir un modelo LLM de HuggingFace y al menos dos modelos de embeddings. Justifique su elección.

### c) Implemente una clase ChatBot usando lo elegido en b).
Puede usar cualquier base de datos vectorial: Chroma y FAISS son las más documentadas. Recuerde que sus datos para su BD conocimiento es el dataset que Ud. planteó en el TP4.

### d) Pruebe el chatbot creado en c) con las preguntas de su dataset a)
Usando los modelos elegidos en b), observe las respuestas generadas por el chatbot comparando al menos dos modelos de embeddings.
Justifique y determine cuál elegiría para su aplicación.

### e) BONUS (opcional)
Evalue el chatbot creado en c) para los modelos de embeddings elegidos y usados en d) usando el dataset a) con las métricas context precision y context recall. Puede probar usar la librería ragas o implementar Ud. mismo el cálculo de dichas métricas.
Dado sus resultados: ¿refuerza o no sus conclusiones realizadas en d) ?
Explique los resultados obtenidos y agregue sus conclusiones.

### f) BONUS (opcional)
Evalue el modelo LLM usado para el chatbot creado en c) con el modelo de embedding que mejor le haya resultado en d) usando el dataset a).
Use las métricas Answer Relevancy y Faithfulness. Puede probar usar la librería ragas o implementar Ud. mismo el cálculo de dichas métricas. Explique los resultados obtenidos y agregue sus conclusiones.

### g) REFERENCIAS.
Es obligatorio citar las fuentes de todo material utilizado para su TP incluido chats con IA generativa. Al final en su apartado Referencias liste todas las fuentes consultadas. Por ejemplo si mantuvo una conversación con ChatGPT puede agregar el enlace compartido a dicha conversación.

### Nota: el codigo anterior debe ejecutarse ya que reutilizo algunas cosas

### Otra nota: este trabajo tarda bastante en terminar de ejecutarse completamente

In [ ]:
!pip install requests==2.32.4
!pip install transformers sentence-transformers faiss-cpu datasets accelerate
!pip install ragas --no-deps
print("Bibliotecas instaladas.")

INFO: pip is looking at multiple versions of langchain-community to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of langchain-openai to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.6/23.6 MB 36.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 366.7/366.7 kB 13.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 160.9/160.9 kB 8.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 43.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 36.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.7/64.7 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.0/76.0 kB 5.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.0/8.0 MB 70.0 MB/

Bibliotecas instaladas.


In [ ]:
import faiss # Para la base de datos vectorial
from transformers import pipeline
from sentence_transformers import SentenceTransformer

# Creamos el DataFrame y la base de conocimiento
df_conocimiento = pd.DataFrame(preguntas_respuestas)
documentos_conocimiento = df_conocimiento["respuestas"].tolist()

print(f"Datos y función de limpieza cargados. {len(documentos_conocimiento)} documentos en la base de conocimiento.")

Datos y función de limpieza cargados. 52 documentos en la base de conocimiento.


## a) Creación del conjunto de datos de evaluación

Para evaluar el chatbot RAG, necesitamos un conjunto de datos de prueba. Este conjunto consistirá en **preguntas nuevas (paráfrasis)** que no están en la base de conocimiento original, pero cuyas respuestas **sí** se encuentran en ella.

Esto nos permite probar si el sistema puede "entender" una pregunta reformulada, encontrar el contexto correcto y generar una respuesta adecuada.

In [ ]:
# Crearemos un diccionario con 'question' (la paráfrasis) y 'ground_truth' (la respuesta original)
data_evaluacion = {
    'question': [
        "¿Qué documentos me piden para ingresar a la carrera?",            # Variación de index 2 (Requisitos)
        "¿Cuál es el valor de la matricula semestral?",                    # Variación de index 5 (Costo)
        "¿En qué turnos se dictan las materias habitualmente?",            # Variación de index 8 (Horarios)
        "Olvidé mi clave del portal, ¿cómo la recupero?",                  # Variación de index 10 (Contraseña)
        "¿En qué fechas se rinden los finales?",                           # Variación de index 14 (Exámenes)
        "Extravié mi identificación de estudiante, ¿qué debo hacer?",      # Variación de index 20 (Pérdida carnet)
        "¿En qué edificio se encuentra la biblioteca central?",            # Variación de index 30 (Ubicación biblioteca)
        "Falté a clase por enfermedad, ¿cómo presento el certificado?",    # Variación de index 41 (Justificar inasistencia)
        "¿Hay algún lugar para comer dentro del campus y dónde queda?",    # Variación de index 43 (Comedor)
        "¿Cuáles son los pasos para solicitar el diploma al terminar?",    # Variación de index 48 (Trámite título)
    ],
    'ground_truth': [
        preguntas_respuestas["respuestas"][2],   # Requisitos admisión
        preguntas_respuestas["respuestas"][5],   # Costo matrícula
        preguntas_respuestas["respuestas"][8],   # Horarios clases
        preguntas_respuestas["respuestas"][10],  # Recuperar contraseña
        preguntas_respuestas["respuestas"][14],  # Exámenes finales
        preguntas_respuestas["respuestas"][20],  # Pérdida carnet
        preguntas_respuestas["respuestas"][30],  # Ubicación biblioteca
        preguntas_respuestas["respuestas"][41],  # Justificar inasistencia
        preguntas_respuestas["respuestas"][43],  # Comedor
        preguntas_respuestas["respuestas"][49]   # Trámite título
    ]
}

# Convertir a DataFrame para mejor visualización
df_evaluacion = pd.DataFrame(data_evaluacion)

print(f"Dataset de evaluación creado con {len(df_evaluacion)} preguntas.")
pd.set_option('display.max_colwidth', None)
df_evaluacion.head(10)

Dataset de evaluación creado con 10 preguntas.


question  \
0          ¿Qué documentos me piden para ingresar a la carrera?   
1                  ¿Cuál es el valor de la matricula semestral?   
2          ¿En qué turnos se dictan las materias habitualmente?   
3                Olvidé mi clave del portal, ¿cómo la recupero?   
4                         ¿En qué fechas se rinden los finales?   
5    Extravié mi identificación de estudiante, ¿qué debo hacer?   
6          ¿En qué edificio se encuentra la biblioteca central?   
7  Falté a clase por enfermedad, ¿cómo presento el certificado?   
8  ¿Hay algún lugar para comer dentro del campus y dónde queda?   
9  ¿Cuáles son los pasos para solicitar el diploma al terminar?   

                                                                                                                                                                                                                                                                                            ground_truth  
0                                                                                                                                                                               Los requisitos incluyen título secundario, examen de ingreso y documentos personales. Visita el sitio web para detalles.  
1                                                                                                                                                                                                        La matrícula varía por carrera, pero promedia $5000 por semestre. Hay descuentos para becarios.  
2                                                                                                                                                                                                El calendario académico está disponible en el portal estudiantil, con fechas de inicio, fin y feriados.  
3                                                                                                                                                                                                                      Ve al portal, selecciona 'Olvidé contraseña' y sigue las instrucciones por email.  
4                                                                                                                                                                                                                          Los exámenes finales son en junio y diciembre, según el calendario académico.  
5                                                                                                                                                                                                                               Reporta la pérdida en seguridad y solicita un duplicado con foto y pago.  
6                                                                                                                                        La biblioteca central está ubicada en el edificio B, planta baja. El horario de atención es de lunes a viernes de 8:00 a 20:00 hs y sábados de 9:00 a 13:00 hs.  
7                                                                                   Para justificar una inasistencia debes presentar el certificado médico o documentación correspondiente en secretaría dentro de las 48 horas. Las inasistencias por razones de fuerza mayor se evalúan caso por caso.  
8                                                                 El comedor universitario está ubicado en el edificio C, primer piso. Ofrece menú estudiantil a precio subsidiado. Horarios: desayuno 7:30-9:00, almuerzo 12:00-15:00, merienda 16:00-18:00. Debes presentar tu credencial estudiantil.  
9  Para tramitar tu título universitario debes: 1) Haber aprobado todas las materias del plan de estudios, 2) Completar el formulario de solicitud de título, 3) Presentar documentación (DNI, partida de nacimiento, certificado analítico), 4) Pagar la tasa de título. El trámite demora 90-120 días.

## b) Elección de Modelos (LLM y Embeddings)

### LLM (Large Language Model)

* **Modelo Elegido:** `Qwen/Qwen2.5-0.5B-Instruct`
* **Justificación:** liviano, rápido y económico, manteniendo buena calidad para tareas básicas de comprensión y generación en español.
    

### Modelos de Embeddings

1.  **Modelo 1 :** `intfloat/multilingual-e5-large`
    * **Justificación:** Optimizado para tareas multilingües con énfasis en más de 100 idiomas, incluyendo el español. Produce embeddings de alta calidad para similitud semántica y es particularmente útil en RAG (Retrieval-Augmented Generation) para chatbots, ofreciendo un buen equilibrio entre tamaño y rendimiento.

2.  **Modelo 2 :** `BAAI/bge-m3`
    * **Justificación:** Un modelo multilingüe que soporta español entre muchos idiomas, con alto rendimiento en embeddings para búsqueda semántica y clustering. Es recomendable por su consistencia en idiomas no ingleses.

## c) Implementación de la Clase ChatBot (RAG)

Crearemos una clase `RAGChatBot` que encapsule toda la lógica:
1.  **`__init__`**: Carga los modelos (LLM y Embeddings) y construye la base de datos vectorial.
2.  **`_build_vector_db`**: Procesa los documentos (respuestas), genera sus embeddings y los almacena en un índice FAISS.
3.  **`retrieve`**: Dada una consulta, genera su embedding y busca los `k` documentos más similares en FAISS.
4.  **`generate_answer`**: Construye el *prompt* (usando la consulta y el contexto recuperado) y se lo pasa al LLM para generar la respuesta final.
5.  **`ask`**: Es el método público que orquesta todo el proceso (retrieve -> generate).

In [ ]:
class RAGChatBot:
    def __init__(self, llm_name, embedding_name, documents):
        print(f"Iniciando RAGChatBot con LLM: {llm_name} y Embeddings: {embedding_name}")

        # 1. Cargar Modelo de Embeddings
        print("Cargando Modelo de Embeddings...")
        self.embedding_model = SentenceTransformer(embedding_name)
        print("Modelo de Embeddings cargado.")

        # 2. Cargar LLM
        print("Cargando LLM...")
        print("  - Cargando modelo generativo...")

        self.generator = pipeline("text-generation",
                                  model=llm_name,
                                  max_length=256,
                                  device=-1)
        print("LLM cargado.")

        # 3. Limpiar y almacenar documentos
        self.documents = documents
        self.documentos_limpios = [limpiar_texto(doc) for doc in self.documents]

        # 4. Construir Base de Datos Vectorial (FAISS)
        print("Construyendo Vector DB (FAISS)...")
        self._build_vector_db()
        print("Vector DB construida exitosamente.")

    def _build_vector_db(self):
        # Generar embeddings para todos los documentos de conocimiento
        embeddings = self.embedding_model.encode(self.documentos_limpios,
                                                 normalize_embeddings=True,
                                                 show_progress_bar=True)

        # Obtener la dimensionalidad de los embeddings
        dimension = embeddings.shape[1]

        # Crear un índice FAISS (IndexFlatL2 mide la distancia L2)
        self.index = faiss.IndexFlatL2(dimension)

        # Agregar los embeddings al índice
        self.index.add(embeddings.astype('float32'))

        print("Sistema RAG listo\n")

    def retrieve(self, query, k=3):
        """Recupera los k documentos más relevantes para una consulta."""
        # Limpiar la consulta
        query_limpia = limpiar_texto(query)

        # Generar embedding para la consulta
        query_embedding = self.embedding_model.encode([query_limpia], normalize_embeddings=True)

        # Buscar en el índice FAISS
        # D = distancias, I = índices
        D, I = self.index.search(query_embedding.astype('float32'), k)

        # Retornar documentos recuperados
        retrieved_docs = []
        for idx, dist in zip(I[0], D[0]):
            retrieved_docs.append({'documento': self.documentos_limpios[idx],
                                   'distancia': float(dist)})

        return retrieved_docs

    def generate_answer(self, query, context):
        """Genera una respuesta usando el LLM basado en la consulta y el contexto."""

        # Crear el prompt para el modelo
        prompt = f"""
Contexto:
{context}

Pregunta:
{query}

Por favor, responde la pregunta basándote únicamente en el contexto proporcionado.

Respuesta:
"""

        # Generar la respuesta
        generated = self.generator(prompt, max_new_tokens=100)[0]['generated_text']
        answer = generated.split("Respuesta:")[-1].strip()

        return answer

    def ask(self, query, k=3, model=""):
        """Orquesta el proceso completo de RAG."""
        # 1. Recuperar contexto
        retrieved_docs = self.retrieve(query, k=k)

        print(f"Pregunta: {query}\n")
        print("Documentos recuperados:")
        for i, doc in enumerate(retrieved_docs, 1): print(f"\n  {i}. [Distancia: {doc['distancia']:.4f}]\n\n {doc['documento']}\n")

        # Unir los documentos en un solo string de contexto
        context_str = "\n---\n".join([doc['documento'] for doc in retrieved_docs])

        # 2. Generar respuesta
        answer = self.generate_answer(query, context_str)

        print(f"modelo {model} Respuesta: {answer}\n")
        print("-" * 80 + "\n")

        # Devolvemos la respuesta y los documentos recuperados (para evaluación)
        return answer, retrieved_docs

In [ ]:
# --- Definición de los modelos a comparar ---
llm_name = "Qwen/Qwen2.5-0.5B-Instruct"
embedding_name_1 = "intfloat/multilingual-e5-large"
embedding_name_2 = "BAAI/bge-m3"

In [ ]:
# --- Instanciación del Chatbot 1 ---
# (Uso los documentos de conocimiento del TP4)
chatbot_1 = RAGChatBot(llm_name=llm_name,
                       embedding_name=embedding_name_1,
                       documents=documentos_conocimiento)

Iniciando RAGChatBot con LLM: Qwen/Qwen2.5-0.5B-Instruct y Embeddings: intfloat/multilingual-e5-large
Cargando Modelo de Embeddings...


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/387 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/690 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.24G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/418 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/280 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/201 [00:00<?, ?B/s]

Modelo de Embeddings cargado.
Cargando LLM...
  - Cargando modelo generativo...


config.json:   0%|          | 0.00/659 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/988M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/242 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cpu


LLM cargado.
Construyendo Vector DB (FAISS)...


Batches:   0%|          | 0/2 [00:00<?, ?it/s]

Sistema RAG listo

Vector DB construida exitosamente.


In [ ]:
# --- Instanciación del Chatbot 2 ---
# (Uso los documentos de conocimiento del TP4)
chatbot_2 = RAGChatBot(llm_name=llm_name,
                       embedding_name=embedding_name_2,
                       documents=documentos_conocimiento)

Iniciando RAGChatBot con LLM: Qwen/Qwen2.5-0.5B-Instruct y Embeddings: BAAI/bge-m3
Cargando Modelo de Embeddings...


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/123 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/54.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/687 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/2.27G [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/2.27G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/444 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.1M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/964 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

Modelo de Embeddings cargado.
Cargando LLM...
  - Cargando modelo generativo...


Device set to use cpu


LLM cargado.
Construyendo Vector DB (FAISS)...


Batches:   0%|          | 0/2 [00:00<?, ?it/s]

Sistema RAG listo

Vector DB construida exitosamente.


## d) Pruebas y Comparación de Chatbots

Ahora probaremos ambos chatbots con el conjunto de datos de evaluación (las preguntas reformuladas). Compararemos las respuestas generadas (`answer`) con la respuesta esperada (`ground_truth`).

In [ ]:
import pandas as pd  # asumo que pandas ya está importado; útil para convertir resultados a DataFrame

# Inicializo una lista donde guardaré cada comparación entre respuesta esperada y respuestas de los chatbots
resultados_comparacion = []

# Recorro cada fila del DataFrame de evaluación (cada fila contiene una pregunta y su ground truth)
for index, row in df_evaluacion.iterrows():
    # Extraigo la pregunta y la respuesta esperada (ground truth) de la fila actual
    pregunta = row['question']
    ground_truth = row['ground_truth']

    # Imprimo por consola para tener trazabilidad de la evaluación en tiempo real
    print(f"--- Evaluando Pregunta: '{pregunta}' ---")

    # Creo un diccionario base para almacenar la comparación de esta pregunta
    resultado_dict = {
        'pregunta': pregunta,
        'respuesta_esperada': ground_truth
    }

    # Consulto al primer chatbot y guardo tanto la respuesta como el contexto devuelto
    answer_1, context_1 = chatbot_1.ask(pregunta, model="chatbot_1")
    resultado_dict['respuesta_bot_1'] = answer_1
    resultado_dict['contexto_bot_1'] = context_1

    # Consulto al segundo chatbot y guardo su respuesta y contexto
    answer_2, context_2 = chatbot_2.ask(pregunta, model="chatbot_2")
    resultado_dict['respuesta_bot_2'] = answer_2
    resultado_dict['contexto_bot_2'] = context_2

    # Muestro la respuesta esperada para facilitar la inspección manual durante la ejecución
    print(f"Respuesta Esperada: {ground_truth}")

    # Añado el diccionario de esta pregunta a la lista de resultados
    resultados_comparacion.append(resultado_dict)

# Convierto la lista de diccionarios a un DataFrame para facilitar el análisis y exportación
df_resultados = pd.DataFrame(resultados_comparacion)

--- Evaluando Pregunta: '¿Qué documentos me piden para ingresar a la carrera?' ---
Pregunta: ¿Qué documentos me piden para ingresar a la carrera?

Documentos recuperados:

  1. [Distancia: 0.2532]

 los requisitos incluyen titulo secundario examen de ingreso y documentos personales visita el sitio web para detalles


  2. [Distancia: 0.2873]

 si puedes cursar hasta 2 materias electivas de otras carreras si hay cupos disponibles y cumplen con tu plan de estudios debes solicitar autorizacion en secretaria academica presentando una nota fundamentando tu pedido


  3. [Distancia: 0.2946]

 solicita becas en la oficina de ayuda financiera con formulario y documentos probatorios

modelo chatbot_1 Respuesta: Los requisitos incluyen un título secundario, un examen de ingreso, y los siguientes documentos: un documento de identificación, una noticia sobre tu solicitud, y dos documentos personales (visitas al sitio web para detalles). Si tienes más de dos materias electivas de otras carreras, pu

### Visualización de Resultados Comparativos

(Mostramos el DataFrame con todas las respuestas)

In [ ]:
# Configurar pandas para ver el texto completo
pd.set_option('display.max_colwidth', None)
display(df_resultados[['pregunta', 'respuesta_bot_1', 'respuesta_bot_2', 'respuesta_esperada']])

pregunta  \
0          ¿Qué documentos me piden para ingresar a la carrera?   
1                  ¿Cuál es el valor de la matricula semestral?   
2          ¿En qué turnos se dictan las materias habitualmente?   
3                Olvidé mi clave del portal, ¿cómo la recupero?   
4                         ¿En qué fechas se rinden los finales?   
5    Extravié mi identificación de estudiante, ¿qué debo hacer?   
6          ¿En qué edificio se encuentra la biblioteca central?   
7  Falté a clase por enfermedad, ¿cómo presento el certificado?   
8  ¿Hay algún lugar para comer dentro del campus y dónde queda?   
9  ¿Cuáles son los pasos para solicitar el diploma al terminar?   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     respuesta_bot_1  \
0                                      Los requisitos incluyen un título secundario, un examen de ingreso, y los siguientes documentos: un documento de identificación, una noticia sobre tu solicitud, y dos documentos personales (visitas al sitio web para detalles). Si tienes más de dos materias electivas de otras carreras, puedes solicitar autorización en secretaría académica. También puede solicitar becas en la Oficina de Aprendizaje Financiero con un formulario y los documentos   
1                                                 5000\n\nExplicación:\nEl contexto indica que la matricula promedio varía según la carrera y se ofrecen descuentos. Sin embargo, no menciona ningún valor específico para el semestre o el precio total de la matrícula semestral. Por lo tanto, el valor más cercano está asociado con las condiciones de pago en la plataforma de pago, específicamente la opción de pago electrónico mediante un código de pago. Este código puede variar depend   
2                     El contexto no proporciona información sobre los turnos por lo que no puedo determinar exactamente en qué turno se dictan las materias habitualmente. Sin embargo, generalmente estas materias se dictan en el primer cuatrimestre del año y el segundo cuatrimestre del mismo, aunque estos se pueden cambiar según sea necesario para satisfacer las necesidades educativas. \n\nSin embargo, esto se debe a una flexibilidad que depende de la estructura educativa actual.   
3                                                   Me aseguro de que mis datos personales se guarden seguros. Por lo tanto, no he compartido ninguna información personal con terceros. Sin embargo, si me envían una notificación o mensaje, estoy seguro de que los ha recibido y de que no me envíaron ningún enlace de recuperación de contraseña. Mi objetivo es mantener mis credenciales confidenciales. Si tienes alguna otra solicitud o necesitas ayuda con algo más, estaré encantado de   
4                                                                junio y diciembre\n\nExplicación:\nEl contexto indica que los examenes finales son en junio y diciembre. Esto se refiere a las fechas específicas por las que se rinden los mismos según el calendario académico. No menciona fechas adicionales o fechas diferentes para los finales. Por lo tanto, la respuesta más precisa es junio y diciembre. \nNo hay otra información relevante que pueda ser relevante para responder esta   
5                                                                                 Debes hacer una denuncia en el campus virtual. Luego, solicitarás la copia de tu cédula de identidad en la Oficina de Alumnos. Deberías pagar la tarifa administrativa correspondiente. Finalmente, tendrías que retirar la nueva cédula de identidad en 5 días hábiles. Por último, deberías completar la denuncia

## Nota: no estoy muy seguro de si varian o cuanto pueden variar los resultados en cada ejecucion del codigo.

## Resultado de los modelos
0. Chatbot 1 = buen resultado │ Chatbot 2 = correcto, pero con agregados que ensucian el texto

1. Chatbot 1 = valor correcto y agrega una explicacion a ese valor │ Chatbot 2 = valor correcto y agrega una explicacion a ese valor

2. Chatbot 1 = no encuentra correlacion, y da como respuesta los inicios de las materias y no los horarios│ Chatbot 2 = no encuentra correlacion, y da como respuesta los inicios de las materias y no los horarios.

ambos modelos relacionaron turnos con fechas de cursada y no con horarios de clase, supongo que esa es la confusión que tuvieron.

3. Chatbot 1 = incorrecta, responde de forma extraña │ Chatbot 2 = me parece correcta, no es conciso como la respuesta esperada pero da un paso a paso de como resolver lo que se pregunta.

4. Chatbot 1 = correcta │ Chatbot 2 = correcta

5. Chatbot 1 = respondio de forma correcta, aunque luego lo mezclo un poco con la perdida de la contraseña │ Chatbot 2 = mal

6. Chatbot 1 =  casi bien, dijo A en lugar de B│ Chatbot 2 = indico bien el edificio, aunque luego incluyo texto en ingles que ensucia la salida

7. Chatbot 1 =  correcto │ Chatbot 2 = correcto

8. Chatbot 1 =  correcto │ Chatbot 2 = correcto

9. Chatbot 1 = se confuncio en el paso a paso │ Chatbot 2 = correcto

# Conclusiones

## Chatbot 1: Embeddings intfloat/multilingual-e5-large
Este modelo presentó un desempeño más estable en cuanto a formato, pero con algunas limitaciones notables en la lógica y la precisión, mayormente sus respuestas fueron gramaticalmente coherentes y completas como en el caso de la Pregunta 5.

El modelo tuvo alucinaciones y algunos fallos de precisión, en la pregunta 6, el modelo alucinó información, indicando erróneamente que la biblioteca estaba en el "Edificio A", cuando el contexto no lo respaldaba ya que estaba en la B, error que me parece bastante grave ya que en el contexto se ve que dice edificio B pero aun asi pone edificio A, tambien, en la pregunta 2, falló en la correlación, confundiendo turnos de horarios con las fechas de calendario, tambien en la pregunta 3 la del olvido de la clave, el modelo respondio de forma extraña, en lugar de responder con más coherencia el procedimiento técnico, luego en la pregunta 9, se confundió en el orden lógico del "paso a paso" para el trámite, resultando en una guía menos útil que la de su competidor.

## Chatbot 2: Embeddings BAAI/bge-m3
Este modelo demostró una capacidad superior en la recuperación de contexto específico, logrando encontrar datos que el primero pasó por alto, pero con errores técnicos críticos en la generación de la respuesta final.

Tuvo en algunas preguntas mayor precisión, ya que a diferencia del primero, este modelo sí encontró el dato correcto sobre la ubicación de la biblioteca el cual era el edificio B en la pregunta 6 y estructuró correctamente los pasos para el obtener el diploma en la pregunta 9.

Tambien sufrió un fallo grave en la pregunta 6, a pesar de tener el dato correcto, inyectó las instrucciones del sistema ("You are an AI assistant...") en la respuesta final, ensuciando totalmente la salida.

Tuvo unos problemas de finalización ya que mostró inestabilidad en la generación de texto, por ejemplo corto abruptamente la respuesta en la pregunta 5 ("6. Retir"), dejando la idea incompleta.

Y tambien mostro la falla en la correlacion de turnos, ya que de igual forma que el primer modelo, falló en la pregunta 2, al no poder responder sobre los horarios, y respondiendo en base a fechas.

## Algunas limitaciones que note
Ambos chatbots fallaron en la pregunta 2 ("¿En qué turnos se dictan las materias...?"), los dos recuperaron información sobre fechas de inscripción (cuatrimestres) en lugar de horarios de cursada (8am a 8pm),
lo cual podria indicar que la palabra "turnos" generó algo de ambigüedad semántica que ninguno de los dos modelos de embeddings pudo resolver perfectamente con el dataset disponible que dentro de todo es pequeño y además, el uso de un LLM pequeño (0.5B parámetros) contribuyó a la inestabilidad gramatical y a las alucinaciones observadas, ya que pruebas con versiones mas grandes que hice con el mismo modelo, me dieron resultados mucho mejores que los que tuve, solo que la limitacion más grande que tuve con esos modelos fue que colab se saturaba muy rapido y cortaba la ejecucion del codigo y tambien aparte el tiempo que tomaban, ya que cada respuesta de esos modelos podia ir de 5 a 20 minutos dependiendo de que tan grande sea el modelo, por lo cual en este caso aunque tuvieron resultados muy buenos resulto practicamente inviable utilizarlos, y tuve que usar este llm que no es muy grande.

## Posibles mejoras
para mejorar el sistema, se podria utilizar un modelo mas grande, aunque eso requiere aumentar el poder de computo, tambien se podria mejorar el promp que se da al modelo aplicando metodos como tokenizer los cuales segun vi pueden ayudar a evitar que el modelo escupa las instrucciones del sistema

# Conclusión Final y Elección
Aunque el resultado de ambos modelos fue algo similar, ya que estuvieron un poco igualado en su desenpeño lo cual complica la eleccion de alguno, ademas de que particularmente no me gusto el desempeño de ninguno, el modelo que usare por ende el ganador es el Chatbot 2 utilizando el embeddings BAAI/bge-m3, ya que aunque sus respuestas están un poco sucias y tuvo algunos fallos, demostró ser más inteligente y capaz de recuperar la información correcta ya que tuvo un poco mas de precision que el modelo 1.

# e) BONUS (opcional)
Context Recall (Exhaustividad): Mide si el sistema fue capaz de encontrar el documento correcto.

Context Precision (Precisión): Penaliza si el documento correcto aparece, pero aparece "abajo" en la lista.

In [ ]:
# =============================================================================
# EVALUACIÓN DE MÉTRICAS: CONTEXT PRECISION Y CONTEXT RECALL (MANUAL)
# =============================================================================

def evaluar_retrieval(chatbot, df_eval, nombre_modelo):
    """
    Calcula Context Recall y Context Precision comparando los documentos recuperados
    con el ground_truth (que sabemos que existe en la base de conocimiento).
    """
    resultados = []
    print(f"Iniciando evaluación para: {nombre_modelo}...")

    for index, row in df_eval.iterrows():
        pregunta = row['question']
        ground_truth = row['ground_truth']

        # 1. Obtenemos los documentos recuperados del chatbot (k=3)
        # Nota: Usamos 'ask' pero solo nos importan los docs recuperados, no la respuesta generada por el LLM
        _, retrieved_docs = chatbot.ask(pregunta, k=3)

        # Extraemos solo el texto de los documentos y lo limpiamos para comparar
        textos_recuperados = [limpiar_texto(doc['documento']) for doc in retrieved_docs]
        ground_truth_limpio = limpiar_texto(ground_truth)

        # --- CÁLCULO DE MÉTRICAS ---

        # 2. Context Recall (Sensibilidad)
        # ¿La respuesta correcta está presente en ALGUNO de los documentos recuperados?
        # 1 = Sí, 0 = No.
        is_retrieved = 1 if ground_truth_limpio in textos_recuperados else 0
        context_recall = is_retrieved

        # 3. Context Precision (Precisión del contexto)
        # Evaluamos qué tan "arriba" apareció el documento correcto.
        # Si está 1ro (rank 1) -> 1.0
        # Si está 2do (rank 2) -> 0.5
        # Si está 3ro (rank 3) -> 0.33
        # Si no está -> 0.0
        context_precision = 0.0
        if is_retrieved:
            rank = textos_recuperados.index(ground_truth_limpio) + 1
            context_precision = 1 / rank

        resultados.append({
            'Modelo': nombre_modelo,
            'Pregunta': pregunta,
            'Context Recall': context_recall,
            'Context Precision': round(context_precision, 4)
        })

    return pd.DataFrame(resultados)

# --- EJECUCIÓN DE LA EVALUACIÓN ---

# 1. Evaluar Chatbot 1 (Embeddings: intfloat/multilingual-e5-large)
df_metrics_1 = evaluar_retrieval(chatbot_1, df_evaluacion, "Chatbot 1 (E5-Large)")

# 2. Evaluar Chatbot 2 (Embeddings: BAAI/bge-m3)
df_metrics_2 = evaluar_retrieval(chatbot_2, df_evaluacion, "Chatbot 2 (BGE-M3)")

# 3. Concatenar y mostrar promedio de resultados
df_final_metrics = pd.concat([df_metrics_1, df_metrics_2], ignore_index=True)

# Calcular los promedios por modelo
resumen_metricas = df_final_metrics.groupby('Modelo')[['Context Recall', 'Context Precision']].mean()

print("\n" + "="*50)
print("RESULTADOS FINALES DE EVALUACIÓN DE EMBEDDINGS")
print("="*50)
display(resumen_metricas)

print("\n" + "="*50)
print("DETALLE POR PREGUNTA")
print("="*50)
display(df_final_metrics)

Iniciando evaluación para: Chatbot 1 (E5-Large)...
Pregunta: ¿Qué documentos me piden para ingresar a la carrera?

Documentos recuperados:

  1. [Distancia: 0.2532]

 los requisitos incluyen titulo secundario examen de ingreso y documentos personales visita el sitio web para detalles


  2. [Distancia: 0.2873]

 si puedes cursar hasta 2 materias electivas de otras carreras si hay cupos disponibles y cumplen con tu plan de estudios debes solicitar autorizacion en secretaria academica presentando una nota fundamentando tu pedido


  3. [Distancia: 0.2946]

 solicita becas en la oficina de ayuda financiera con formulario y documentos probatorios

modelo  Respuesta: Título secundario examen de ingreso y documentos personales. 

La respuesta es que necesito un título secundario examen de ingreso junto con documentos personales para ingresar a la carrera. Además, puedo cursar hasta dos materias electivas de otras carreras si tengo disponibles los cupos y cumeto con mi plan de estudio. Para s

Context Recall  Context Precision
Modelo                                                 
Chatbot 1 (E5-Large)             0.8            0.75000
Chatbot 2 (BGE-M3)               1.0            0.81666


DETALLE POR PREGUNTA


Modelo  \
0   Chatbot 1 (E5-Large)   
1   Chatbot 1 (E5-Large)   
2   Chatbot 1 (E5-Large)   
3   Chatbot 1 (E5-Large)   
4   Chatbot 1 (E5-Large)   
5   Chatbot 1 (E5-Large)   
6   Chatbot 1 (E5-Large)   
7   Chatbot 1 (E5-Large)   
8   Chatbot 1 (E5-Large)   
9   Chatbot 1 (E5-Large)   
10    Chatbot 2 (BGE-M3)   
11    Chatbot 2 (BGE-M3)   
12    Chatbot 2 (BGE-M3)   
13    Chatbot 2 (BGE-M3)   
14    Chatbot 2 (BGE-M3)   
15    Chatbot 2 (BGE-M3)   
16    Chatbot 2 (BGE-M3)   
17    Chatbot 2 (BGE-M3)   
18    Chatbot 2 (BGE-M3)   
19    Chatbot 2 (BGE-M3)   

                                                        Pregunta  \
0           ¿Qué documentos me piden para ingresar a la carrera?   
1                   ¿Cuál es el valor de la matricula semestral?   
2           ¿En qué turnos se dictan las materias habitualmente?   
3                 Olvidé mi clave del portal, ¿cómo la recupero?   
4                          ¿En qué fechas se rinden los finales?   
5     Extravié mi identificación de estudiante, ¿qué debo hacer?   
6           ¿En qué edificio se encuentra la biblioteca central?   
7   Falté a clase por enfermedad, ¿cómo presento el certificado?   
8   ¿Hay algún lugar para comer dentro del campus y dónde queda?   
9   ¿Cuáles son los pasos para solicitar el diploma al terminar?   
10          ¿Qué documentos me piden para ingresar a la carrera?   
11                  ¿Cuál es el valor de la matricula semestral?   
12          ¿En qué turnos se dictan las materias habitualmente?   
13                Olvidé mi clave del portal, ¿cómo la recupero?   
14                         ¿En qué fechas se rinden los finales?   
15    Extravié mi identificación de estudiante, ¿qué debo hacer?   
16          ¿En qué edificio se encuentra la biblioteca central?   
17  Falté a clase por enfermedad, ¿cómo presento el certificado?   
18  ¿Hay algún lugar para comer dentro del campus y dónde queda?   
19  ¿Cuáles son los pasos para solicitar el diploma al terminar?   

    Context Recall  Context Precision  
0                1             1.0000  
1                1             1.0000  
2                0             0.0000  
3                1             1.0000  
4                1             1.0000  
5                0             0.0000  
6                1             1.0000  
7                1             1.0000  
8                1             1.0000  
9                1             0.5000  
10               1             1.0000  
11               1             1.0000  
12               1             0.3333  
13               1             1.0000  
14               1             1.0000  
15               1             0.3333  
16               1             1.0000  
17               1             1.0000  
18               1             1.0000  
19               1             0.5000

## Context Recall
El modelo Chatbot 2 alcanzó un Context Recall perfecto de 1.0, lo que significa que recuperó la información relevante en el 100% de las preguntas mientras que por el contrario, el Chatbot 1 obtuvo un 0.8, fallando en recuperar el contexto en el 20% de los casos.

## Context Precision
El Chatbot 2 también lidera en Context Precision con un 0.816 vs 0.750, lo cual podria indicar que los fragmentos de texto que el Chatbot 2 entrega al LLM son más relevantes y contienen menos ruido o información innecesaria que los entregados por el Chatbot 1.

### Conclusion
Estos datos refuerzan el uso del modelo 2 ya que mientras que el Chatbot 1 falla estructuralmente en encontrar la información en 2 de cada 10 consultas, el Chatbot 2 demuestra ser un RAG mucho más robusto y fiable, garantizando que el LLM siempre tenga acceso a los datos necesarios para responder.

# f) BONUS (opcional)

modelo de embeddings ganador (chatbot_2 con BAAI/bge-m3):

Answer Relevancy (Proxy): Calcularemos la Similitud Coseno entre el vector de la Pregunta y el vector de la Respuesta Generada. Si la respuesta responde lo que se preguntó, deben estar cerca semánticamente.

Faithfulness (Proxy): Calcularemos la Similitud Coseno entre el vector de la Respuesta Generada y el vector del Contexto Recuperado. Una alta similitud indica que la respuesta se adhiere (es fiel) a la información encontrada y no es una alucinación externa.

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

def evaluar_generacion_llm(chatbot, df_eval):
    """
    Evalúa la calidad de la generación del LLM usando métricas vectoriales.
    Utiliza el modelo de embeddings del propio chatbot para medir distancias semánticas.
    """
    resultados = []
    print(f"Iniciando evaluación de generación con el modelo de embeddings: {chatbot.embedding_model.modules}")

    for index, row in df_eval.iterrows():
        pregunta = row['question']

        # 1. Obtener respuesta y contexto del chatbot
        # (Usamos k=3 porque fue tu configuración estándar)
        respuesta_generada, docs_recuperados = chatbot.ask(pregunta, k=3, model="Evaluación")

        # Unimos el texto del contexto recuperado para vectorizarlo
        contexto_unido = " ".join([d['documento'] for d in docs_recuperados])

        # 2. Generar Embeddings (Vectorización)
        # Necesitamos vectores para: Pregunta, Respuesta y Contexto
        emb_pregunta = chatbot.embedding_model.encode([pregunta])
        emb_respuesta = chatbot.embedding_model.encode([respuesta_generada])
        emb_contexto = chatbot.embedding_model.encode([contexto_unido])

        # 3. Calcular Métricas (Similitud del Coseno)

        # A) Answer Relevancy: ¿Qué tanto se relaciona la respuesta con la pregunta?
        relevancy_score = cosine_similarity(emb_pregunta, emb_respuesta)[0][0]

        # B) Faithfulness: ¿Qué tanto se relaciona la respuesta con el contexto recuperado?
        # (Si el score es bajo, significa que el modelo generó texto ajeno al contexto)
        faithfulness_score = cosine_similarity(emb_respuesta, emb_contexto)[0][0]

        resultados.append({
            'Pregunta': pregunta,
            'Respuesta Generada': respuesta_generada,
            'Answer Relevancy': round(relevancy_score, 4),
            'Faithfulness': round(faithfulness_score, 4)
        })

    return pd.DataFrame(resultados)

# --- EJECUCIÓN ---

# Usamos chatbot_2 porque determinaste en el punto d) que era el mejor
print("Evaluando LLM del Chatbot 2 (Ganador)...")
df_resultados_llm = evaluar_generacion_llm(chatbot_2, df_evaluacion)

# Calcular promedios
promedio_relevancia = df_resultados_llm['Answer Relevancy'].mean()
promedio_fidelidad = df_resultados_llm['Faithfulness'].mean()

print("\n" + "="*60)
print(f"RESULTADOS FINALES DE GENERACIÓN (LLM + Embeddings BAAI/bge-m3)")
print("="*60)
print(f"Promedio Answer Relevancy: {promedio_relevancia:.4f}")
print(f"Promedio Faithfulness:     {promedio_fidelidad:.4f}")

print("\n" + "="*60)
print("DETALLE POR PREGUNTA")
print("="*60)
# Mostramos solo las columnas clave para facilitar la lectura
display(df_resultados_llm[['Pregunta', 'Answer Relevancy', 'Faithfulness']])

Evaluando LLM del Chatbot 2 (Ganador)...
Iniciando evaluación de generación con el modelo de embeddings: <bound method Module.modules of SentenceTransformer(
  (0): Transformer({'max_seq_length': 8192, 'do_lower_case': False, 'architecture': 'XLMRobertaModel'})
  (1): Pooling({'word_embedding_dimension': 1024, 'pooling_mode_cls_token': True, 'pooling_mode_mean_tokens': False, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
)>
Pregunta: ¿Qué documentos me piden para ingresar a la carrera?

Documentos recuperados:

  1. [Distancia: 0.6711]

 los requisitos incluyen titulo secundario examen de ingreso y documentos personales visita el sitio web para detalles


  2. [Distancia: 0.7213]

 si presenta certificados y solicita convalidacion en la secretaria academica


  3. [Distancia: 0.7397]

 para inscribirte a una materia necesitas 1 tener ap

Pregunta  \
0          ¿Qué documentos me piden para ingresar a la carrera?   
1                  ¿Cuál es el valor de la matricula semestral?   
2          ¿En qué turnos se dictan las materias habitualmente?   
3                Olvidé mi clave del portal, ¿cómo la recupero?   
4                         ¿En qué fechas se rinden los finales?   
5    Extravié mi identificación de estudiante, ¿qué debo hacer?   
6          ¿En qué edificio se encuentra la biblioteca central?   
7  Falté a clase por enfermedad, ¿cómo presento el certificado?   
8  ¿Hay algún lugar para comer dentro del campus y dónde queda?   
9  ¿Cuáles son los pasos para solicitar el diploma al terminar?   

   Answer Relevancy  Faithfulness  
0            0.6258        0.8699  
1            0.6456        0.8153  
2            0.4682        0.7017  
3            0.6601        0.7218  
4            0.5926        0.8081  
5            0.6432        0.8802  
6            0.4550        0.7939  
7            0.7125        0.8494  
8            0.6251        0.9455  
9            0.7336        0.8691

## Faithfulness
El promedio de 0.8255 indica que el modelo tiene una fuerte tendencia a adherirse al contexto proporcionado y no inventar información externa.

## Answer Relevancy
El puntaje de relevancia es notablemente más bajo 0.62 aproximadamente, lo cual podria confirmar el evidente problema de la suciedad en las respuestas.

### Conclusion
Estos datos pueden indicar que el chatbot 2 es un sistema honesto y desprolijo, ya que su alta fidelidad garantiza que la información base sea correcta, pero su moderada relevancia puede indicar que requiere una capa de limpieza quizas para filtrar las explicaciones internas y las fugas de instrucciones antes de enviar la respuesta.

# Referencias
notebook que use para armar la clase rag:

**https://colab.research.google.com/drive/1Z5JfUZNO34fC5VzyUF3P_fQiDLcdmTOb?usp=sharing**

tambien utilice los videos de las clases adjuntas en la clase 12.

uso del modelo gemini para generar la base de datos para el punto a y testear los modelos:

**https://gemini.google.com/share/4b5f89ca58d1**

uso grok para tratar de corregir los indices no correctos para el testeo:

**https://grok.com/share/c2hhcmQtMi1jb3B5_4b584a4e-b736-4eaa-bf2d-139303c32ab1**

nota: no funciono muy bien, asi que los cambie manualmente.

charla con el modelo de IA grok para seleccionar los modelos llm y embeddings:

**https://grok.com/share/c2hhcmQtMi1jb3B5_72d7fdb5-6b58-4021-9e66-7574feffcd4d**

charla con el modelo GPT sobre una advertencia que note:

**https://chatgpt.com/share/691b6d32-3aac-8000-acd2-2f19d2e75763**

charla con el modelo GPT, sobre problemas de rendimiento en el modelo llm:

**https://chatgpt.com/share/691b7ce2-dbc0-8000-bec5-6e07005d3e7d**

**https://chatgpt.com/share/691b7d0b-f324-8000-abfb-959f366fb37a**

charla con el modelo GPT para modificar un par de parametros en la clase rag:

**https://chatgpt.com/share/691b8add-0d3c-8000-ace4-e293b0d08cd1**

otro error que tuve durante la ejecucion:

**https://chatgpt.com/share/691b959e-06f8-8000-a16a-d67845a8b9ae**

al final como el modelo llm que elegi al comienzo era demasiado lento para cargar y dar las respuestas, le pedi a GPT que me recomiende alternativas:

**https://chatgpt.com/share/691b9603-b638-8000-8607-cc18ce9adac9**

charla con GPT sobre el rendimiento del llm y versiones mas ligeras para usar:

**https://chatgpt.com/share/691bc69c-030c-8000-a501-583e538ab365**

chat con el modelo de gemini para hacer los calculos de los dos bonus:

**https://gemini.google.com/share/788e44ee079f**

consulta a geminis sobre el resultado del primer bonus:

**https://gemini.google.com/share/15772afb6e99**

consulta a geminis sobre el resultado del segundo bonus:

**https://gemini.google.com/share/584eaa9e544c**

documentacion oficial de la libreria utilizada para la carga y manejo del LLM:

**https://huggingface.co/docs/transformers/index**

framework utilizado para generar los embeddings de las preguntas y respuestas:

**https://www.sbert.net/**

libreria utilizada para la indexacion y busqueda eficiente de vectores de similitud.

**https://github.com/facebookresearch/faiss**